### DISCUSSION TITLE
Posix and mandatory persistence

### DISCUSSION THREAD

**Discussion by guptask** [facebook/CacheLib#69] [Q&A] [answered]
In the current code, PersistenceManager can only be used with Posix (https://github.com/facebook/CacheLib/blob/main/cachelib/persistence/PersistenceManager.h#L116).
From the configuration side, both POSIX SHM and Sys-V SHM can be configured with persistence. However, we cannot use POSIX SHM without persistence (refer to description of usePosixForShm - throw std::invalid_argument if called without enabling cachePersistence() ). Why does Posix SHM require mandatory persistence ?

**Comment by leozzx** [comment, marked as answer]
Hi @guptask, thanks for reaching out.
Can I know the reason why do you need to set usePosixForShm without enabling persistence?
The reason we require mandatory persistence is because usePosixForShm option is not useful without persistence. The usePosixForShm is only used to create ShmManager for persistence purpose.

**Reply by sathyaphoenix** [reply]
@guptask There are two persistence primitives here and I want to make sure that we are not confusing them. The Shm based persistence that is available through CacheAllocator constructor (SharedMemAttach/SharedMemNew) are for same-host process restart persistence of the cache. This is what kicks in when you have an application restart their binary and want to preserve the cache on the same host. The [PersistenceManager](https://github.com/facebook/CacheLib/blob/main/cachelib/persistence/PersistenceManager.h#L116) you refer to in your question is for x-host persistence where we want to move a cache from one physical host to another. In such scenarios, we only support the posix shm. The reason we don't support Sys-V there  is purely for internal implementation details (we can if we need to).

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
